[NUI] XAML guide (#1007)
Samsung/tizen-docs: docs/application/dotnet/guides/nui/xaml/xaml-support-for-tizen-nui.md

using Tizen.NUI;
using Tizen.NUI.BaseComponents;

namespace NUIXamlTemplate
{
    class Program : NUIApplication
    {
        Animation animation;
        protected override void OnCreate()
        {
            base.OnCreate();
            Window.Instance.BackgroundColor = Color.Blue;
            Window.Instance.KeyEvent += OnKeyEvent;

            XamlPage page = new XamlPage(Window.Instance);

            animation = new Animation(2000);
            animation.AnimateTo(page.text, "Orientation", new Rotation(new Radian(new Degree(180.0f)), PositionAxis.X), 0, 500);
            animation.AnimateTo(page.text, "Orientation", new Rotation(new Radian(new Degree(0.0f)), PositionAxis.X), 500, 1000);
            animation.Looping = true;
            animation.Play();
        }

        public void OnKeyEvent(object sender, Window.KeyEventArgs e)
        {
            if (e.Key.State == Key.StateType.Down && (e.Key.KeyPressedName == "XF86Back" || e.Key.KeyPressedName == "Escape"))
            {
                Exit();
            }
        }

        static void Main(string[] args)
        {
            var app = new Program();
            app.Run(args);
        }
    }
}
```

> When you compile and run this program, the `TextLabel` element will appear on the window. Also you can see the text of "Hello, NUI XAML APP", which is defined in xaml file, animates rotating on the x-axis.

Reviewer: Suggest rephrasing:

When you compile and run this program, the `TextLabel` element appears on the window. Additionally, you can see the "Hello, NUI XAML APP" text, which is defined in the XAML file that rotates on the x-axis.

Question: "animates rotating on the x-axis", this part is not clear. Please verify if the rephrased content is technically fine. Else, please provide more explanation to rephrase better.

> ## Create Solution Before Tizen 5.5

Reviewer: Suggestion: Could we give another heading?
 For example: Create Solution for Tizen Version 5.0 and Lower

@mijin85cho, please suggest.

Reviewer 2: @surbhig12 I understand you tried to keep the consistency for the words, but how about using shorter words for the **title**?
For example like :

Create Solution Under Tizen 5.5

Reviewer: @mijin85cho,
In that case, there is a confusion whether version 5.5 is included or not. To be more clear, I suggested "Tizen Version 5.0 and Lower".

However, I agree that the heading is a bit long but could we take this as an exception considering there would be no confusion with respect to versions?
Please correct me if I am wrong.

Reviewer 2: @surbhig12 Thanks for sharing your concern.

I thought the "under" has the meaning of "less than" as like below link. So I used "under" as it is clear that Tizen 5.5 is not included in my thought.

https://dictionary.cambridge.org/ko/%EC%82%AC%EC%A0%84/%EC%98%81%EC%96%B4/under

![image](https://user-images.githubusercontent.com/9958374/73634167-e7281a00-46a3-11ea-9eca-8fd9e36f1418.png)

However as "Create Solution Under Tizen 5.5" is not tidy, the other form, you suggested can be considered. You are correct that I don't think long sentence is proper to the title, but for the clear meaning, we may can give up the simplicity.

Before we go with your suggestion, if you can give us another suggestion, please let me know.

Reviewer: @mijin85cho, thank ou for the discussion.
The only best option to avoid confusion and ensure clear meaning  I can suggest:

Create Solution Using Tizen Version 5.0 and Lower

# XAML Support for Tizen.NUI

NUI adopts and integrates Xamarin.Forms XAML as NUI XAML. NUI XAML supports almost all the features of Xamarin.Forms XAML along with some additional features. 

> In this file the additional features are explained. However, to get the basic concept and insight of XAML, it is strongly recommended to go through the [eXtensible Application Markup Language (XAML)](/docs.microsoft.com/en-us/xamarin/xamarin-forms/xaml/) guide.

Reviewer: Please add a comma after "In this file"

```suggestion
In this file, the additional features are explained. However, to get a basic concept and insight of XAML, it is strongly recommended to go through the [eXtensible Application Markup Language (XAML)](/docs.microsoft.com/en-us/xamarin/xamarin-forms/xaml/) guide.
```

- Update the UI and the business logic layer separately as they are not coupled.
- Set different layout-xaml files for different configurations, for example, resolution 1080 and 720. When the resolution is changed, the display layout is changed automatically.

## Create Solution From Tizen 5.5

You can create a new Tizen.NUI solution using Visual Studio and then you can edit your first XAML file. For more information, see [Part 1. Getting Started with XAML](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/xaml/xaml-basics/get-started-with-xaml?tabs=windows) in the XAML guide.

To create a new project, follow these steps:

1. In Visual Studio, select **File > New > Project**.

2. In the **Create a new project** dialog that appears, select **C#** from the **All languages** drop-down list.

3. In the **All project types** drop-down list, select **Tizen 5.5**.

4. Select **Tizen NUI Xaml App** and click **Next**.

> **Note**
>
> If you create a new project with **Tizen NUI Xaml App** template, a XAML file and a C# code-behind file are created.

![CreateNUIProject](./media/nui_xaml_template.png)

### Create XAML file

> After creating the **NUIXamlTemplate** solution, you can see `XamlPage.xaml` in `/res/layout` folder. 

Reviewer: Please add article "the":

```suggestion
After creating the **NUIXamlTemplate** solution, you can see `XamlPage.xaml` in the `/res/layout` folder.
```

            animation = new Animation(2000);
            animation.AnimateTo(page.text, "Orientation", new Rotation(new Radian(new Degree(180.0f)), PositionAxis.X), 0, 500);
            animation.AnimateTo(page.text, "Orientation", new Rotation(new Radian(new Degree(0.0f)), PositionAxis.X), 500, 1000);
            animation.Looping = true;
            animation.Play();
        }

        public void OnKeyEvent(object sender, Window.KeyEventArgs e)
        {
            if (e.Key.State == Key.StateType.Down && (e.Key.KeyPressedName == "XF86Back" || e.Key.KeyPressedName == "Escape"))
            {
                Exit();
            }
        }

        static void Main(string[] args)
        {
            var app = new Program();
            app.Run(args);
        }
    }
}
```

When you compile and run this program, the `TextLabel` element appears on the window. Additionally, you can see the "Hello, NUI XAML APP" text, which is defined in the XAML file that rotates on the x-axis.

## Create Solution for Tizen Version 5.0 and Lower

To begin editing your first XAML file, use Visual Studio to create a new Tizen.NUI solution.

> In Visual Studio, select **File > New > Project** from the menu. In the **Create a new project** dialog that appears, select **C# > Tizen** from the selecting box, and then select **Blank App (Tizen.NUI)**.

Reviewer: Please use numbered list and please verify:

To create a new project, follow these steps:

1. In Visual Studio, select **File > New > Project**.
2. In the **Create a new project** dialog that appears, select **C#** from the **All languages** drop-down list.
3. In the **All project types** drop-down list, select **Tizen**.
4. Select **Blank App (Tizen.NUI)** and click **Next**.

The `BindingContext` property is one of the two ways to link the source and target objects. You can include a reference to the source object within the binding expression.
The following code shows how the source object and source property can be specified in the Binding markup extension:

```xml
<TextLabel x:Name="label" Name ="label" Text="Text" Position2D="100,100" Size2D="300,50" PositionX="{Binding Source={x:Reference Name=slider}, Path=Value}" />

<Slider x:Name="slider" Name="slider" LowerBound="100" UpperBound="800" Value="100" ShowPopup="true" ShowValue="false" Position2D="400,400" Size2D="300,20" />
```

The `Binding` markup extension has two arguments, one of which is a markup extension for `x:Reference`. Therefore, a pair of curly braces are nested within the main curly braces:

```xml
PositionX="{Binding Source={x:Reference Name=slider}, Path=Value}"
```

Following are the two ways to specify the link between the source object with the target object:
- Use the `BindingContext` to reference the source object.
- Use the `Source` property of the `Binding` markup extension.

If you specify both, the `Source` property takes precedence over the `BindingContext`.

> For more information, please see https://docs.microsoft.com/en-US/xamarin/xamarin-forms/xaml/xaml-basics/data-binding-basics.

Reviewer: Please remove the word "please":

```suggestion
For more information, see https://docs.microsoft.com/en-US/xamarin/xamarin-forms/xaml/xaml-basics/data-binding-basics.
```

# XAML Support for Tizen.NUI

NUI adopts and integrates Xamarin.Forms XAML as NUI XAML. NUI XAML supports almost all the features of Xamarin.Forms XAML along with some additional features. 
In this file, the additional features are explained. However, to get a basic concept and insight of XAML, it is strongly recommended to go through the [eXtensible Application Markup Language (XAML)](https://docs.microsoft.com/en-US/xamarin/xamarin-forms/xaml/) guide.

## Benefits of XAML

Following are some of the XAML benefits. You can use XAML to:

- Update the UI without compiling and deploying the application.
- Update the UI and the business logic layer separately as they are not coupled.
- Set different layout-xaml files for different configurations, for example, resolution 1080 and 720. When the resolution is changed, the display layout is changed automatically.

> ## Create Solution From Tizen 5.5

Reviewer: ## Create Solution Using Tizen Version 5.5

For consistency, please update.

4. Select **Tizen NUI Xaml App** and click **Next**.

> **Note**
>
> If you create a new project with **Tizen NUI Xaml App** template, a XAML file and a C# code-behind file are created.

![CreateNUIProject](./media/nui_xaml_template.png)

### Create XAML file

After creating the **NUIXamlTemplate** solution, you can see `XamlPage.xaml` in the `/res/layout` folder. 

![SetProjectName](./media/project_name.PNG)

> **Note**
>
> If you select other templates or work on lower version than Tizen 5.5, you have to manually create a XAML file.

If your application runs on a target having different resolutions, such as 720x1080, 480x800, or both of them then put XAML files into the corresponding folders. For example, you can create two folders: `/res/layout/720x1080/` and `/res/layout/480x800/`. The former is the folder with a XAML file for 720x1080 resolution, and the latter is the folder with a XAML file for 480x800 resolution.

In the following sample, `XamlPage.xaml` is created and placed in the **/res/layout/** folder:

![XAMLFilePath](./media/solution_explorer.PNG)

To add any control you want, edit the XamlPage.xaml file:

![XAMLFile](./media/xaml-page-xaml-file.PNG)

> In this sample, the `TextLabel` element is added, and the `Text`, `PointSize`, `TextColor` and other properties are set.

Reviewer: Please add a serial comma:

```suggestion
In this sample, the `TextLabel` element is added, and the `Text`, `PointSize`, `TextColor`, and other properties are set.
```

  <ContentPage.XamlResources>
    <ResourceDictionary>
      <x:String x:Key="urlKey">
          *Resource*/res/picture_m_1.jpg
      </x:String>
      <x:String x:Key="positionKey">
          200, 200
      </x:String>
    </ResourceDictionary>
  </ContentPage.XamlResources>`

  <ImageView x:Name="img1" ResourceUrl="{StaticResource urlKey}"  Position2D="{DynamicResource positionKey}/>
</ContentPage>
```

`StaticResource` accesses the item in the dictionary only once, while XAML is being parsed and the page is being built. However, `DynamicResource` maintains a link between the dictionary key and the property set from that dictionary item. If the item in the resource dictionary referenced by the key changes, then `DynamicResource` will detect that change and set the new value to the property.
When you change the value of `"positionKey"`, `DynamicResource` will detect that change and set the new value to the `Position2D` property.

```csharp
Tizen.NUI.Binding.ResourceDictionary dict = Tizen.NUI.GetResourcesProvider.Get().Resources;
Tizen.NUI.GetResourcesProvider.Get().Resources["positionKey"] = positionX.ToString() + "," + positionY.ToString();
```

> For more information, please see https://docs.microsoft.com/en-US/xamarin/xamarin-forms/xaml/resource-dictionaries.

Reviewer: Please remove the word "please":

```suggestion
For more information, see https://docs.microsoft.com/en-US/xamarin/xamarin-forms/xaml/resource-dictionaries.
```

## TypeConverters

If you want to convert a string to some custom type in XAML, you need to define a TypeConverter. This changes the string format as a XAML format such as converting `Red` to `Color.Red`.
If you write a custom class, and you want instances of your class to be usable as XAML settable attribute values, you might need to write a custom TypeConverter class.

> **Note**
> 

> > NUI XAML is very similar to WPF XAML. For more information on basic concepts, please see

Reviewer: Please remove the word "please":

```suggestion
> NUI XAML is very similar to WPF XAML. For more information on basic concepts, see https://docs.microsoft.com/en-us/dotnet/framework/wpf/advanced/typeconverters-and-xaml.
```
